File: built-in plugins dsp/mb-bass-sub.cpp
/**
 * MB Sub Bass
 * Category : instrument
 * Type     : bass
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Deep sub bass
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_BASS_SUB_H
#define MB_BASS_SUB_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbBassSub : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-bass-sub";
    static constexpr const char* PLUGIN_NAME    = "MB Sub Bass";
    static constexpr const char* PLUGIN_TYPE    = "bass";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float sub = 1.0f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbBassSub() = default;
    ~MbBassSub() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.sub = std::clamp(params.sub, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Sub Bass
        return input;
    }
};

#endif // MB_BASS_SUB_H

File: built-in plugins dsp/mb-bass-sub.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbBassSubPlugin: PluginDefinition = {
    id: 'mb-bass-sub', slug: 'mb-bass-sub', name: 'MB Sub Bass', category: 'instrument', type: 'bass', version: '1.0.0',
    description: 'Deep sub bass', author: 'Max Booster', grade: 'A',
    oscillators: [{ type: 'sine', detune: 0, gain: 1.0 }],
    envelope: { attack: 0.01, decay: 0.2, sustain: 0.8, release: 0.2 },
    parameters: [
      { id: 'sub', name: 'Sub', type: 'float', defaultValue: 1.0, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { sub: 1.0, volume: 0.8 },
  };

export default MbBassSubPlugin;
